File: project/app/src/main/java/org/owntracks/android/ui/contacts/ContactsActivity.java
package org.owntracks.android.ui.contacts;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;

import org.owntracks.android.R;
import org.owntracks.android.databinding.UiContactsBinding;
import org.owntracks.android.geocoding.GeocoderProvider;
import org.owntracks.android.model.FusedContact;
import org.owntracks.android.ui.base.BaseActivity;
import org.owntracks.android.ui.base.BaseAdapter;

import javax.inject.Inject;


public class ContactsActivity extends BaseActivity<UiContactsBinding, ContactsMvvm.ViewModel> implements ContactsMvvm.View, BaseAdapter.ClickListener<FusedContact> {
    private final ContactsAdapter contactsAdapter = new ContactsAdapter(this);

    @Inject
    GeocoderProvider geocoderProvider;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasEventBus(false);
        bindAndAttachContentView(R.layout.ui_contacts, savedInstanceState);
        setSupportToolbar(binding.toolbar);
        setDrawer(binding.toolbar);
        binding.getVm().getContacts().observe(this, contacts -> {
            contactsAdapter.setContactList(contacts.values());
            for (FusedContact contact : contacts.values()) {
                contact.getMessageLocation().removeObservers(binding.getLifecycleOwner());
                contact.getMessageLocation().observe(binding.getLifecycleOwner(), messageLocation -> geocoderProvider.resolve(messageLocation));
            }
        });
        binding.recyclerView.setLayoutManager(new LinearLayoutManager(this));
        binding.recyclerView.setAdapter(contactsAdapter);
    }

    @Override
    public void onClick(@NonNull FusedContact object, @NonNull View view, boolean longClick) {
        viewModel.onContactClick(object);
    }
}

File: project/app/src/main/java/org/owntracks/android/ui/contacts/ContactsActivity.kt
package org.owntracks.android.ui.contacts

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.LinearLayoutManager
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject
import org.owntracks.android.R
import org.owntracks.android.databinding.UiContactsBinding
import org.owntracks.android.model.FusedContact
import org.owntracks.android.support.DrawerProvider
import org.owntracks.android.ui.map.MapActivity

@AndroidEntryPoint
class ContactsActivity :
    AppCompatActivity(),
    AdapterClickListener<FusedContact> {
    @Inject
    lateinit var drawerProvider: DrawerProvider

    private val vm: ContactsViewModel by viewModels()
    private lateinit var contactsAdapter: ContactsAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        contactsAdapter = ContactsAdapter(this, vm.coroutineScope)
        DataBindingUtil.setContentView<UiContactsBinding>(this, R.layout.ui_contacts)
            .apply {
                appbar.toolbar.run {
                    setSupportActionBar(this)
                    drawerProvider.attach(this)
                }
                contactsRecyclerView.run {
                    layoutManager = LinearLayoutManager(this@ContactsActivity)
                    adapter = contactsAdapter
                }
            }
        vm.contacts.observe(this) { contacts: Map<String, FusedContact> ->
            contactsAdapter.setContactList(contacts.values)
            vm.refreshGeocodes()
        }
    }

    override fun onClick(item: FusedContact, view: View, longClick: Boolean) {
        startActivity(
            Intent(this, MapActivity::class.java).putExtra(
                "_args",
                Bundle().apply { putString(MapActivity.BUNDLE_KEY_CONTACT_ID, item.id) }
            )
        )
    }
}
